### PR TITLE
Request & follow redirects for /media/v3/download

### LINKED ISSUE #15196: Synapse calls `/r0` for media over federation
https://github.com/matrix-org/synapse/blob/4fc8875876374ec8f97a3b3cc344a4e3abcf769f/synapse/media/media_repository.py#L427

It should be trying the `/v3` endpoint until something like [MSC1902](https://github.com/matrix-org/matrix-spec-proposals/pull/1902) or similar lands. Arguably there should also be a fallback mechanism to try `/r0` when `/v3` fails, per `M_UNRECOGNIZED` in the [standard error codes](https://spec.matrix.org/v1.6/client-server-api/#common-error-codes).

### PR DESCRIPTION
**Pull request by clokep**
This has two mostly unrelated changes to it:

* Request `/media/v3/download` over federation (falling back to `/media/r0/download`)
* When using the `v3` endpoint, add `allow_redirects` from [MSC3860](https://github.com/matrix-org/matrix-spec-proposals/pull/3860)

I tested this by setting up a server locally and doing:

```
curl --header "Authorization: Bearer $TOK" http://localhost:8080/_matrix/media/v3/download/beeper.com/18850ea089e0ecc16d7db55527925b43ad63295c
```

Which had the following logs:

```
synapse.http.matrixfederationclient - 665 - DEBUG - GET-0 - {GET-O-1} [beeper.com] Sending request: GET matrix-federation://beeper.com/_matrix/media/v3/download/beeper.com/18850ea089e0ecc16d7db55527925b43ad63295c?allow_remote=false&timeout_ms=20000&allow_redirect=true; timeout 60.000000s
synapse.http.matrixfederationclient - 665 - DEBUG - GET-0 - {GET-O-2} [beeper.com] Sending request: GET https://thumbnails.eu-central-2.media.beeper.com/8ca/7d9a66bed41b67440ab3807a03e70e22ac70b; timeout 60.000000s
synapse.http.matrixfederationclient - 716 - DEBUG - GET-0 - {GET-O-2} [beeper.com] Got response headers: 200 OK
synapse.http.matrixfederationclient - 1491 - INFO - GET-0 - {GET-O-1} [beeper.com] Completed: 200 OK [140970 bytes] GET matrix-federation://beeper.com/_matrix/media/v3/download/beeper.com/18850ea089e0ecc16d7db55527925b43ad63295c?allow_remote=false&timeout_ms=20000&allow_redirect=true
```

You can see it attempts to fetch the media from Beeper's homeserver, gets a redirect to a CDN and then fetches it from there instead.

In order to accomplish this I reworked the fetching of downloaded media to be more similar to other federation requests. You can compare the solid to the dotted line. The main goal of this was to add the fallback from v3 -> r0.

```mermaid
flowchart
    A[MediaRepository._download_remote_file]
    B[FederationClient.download_media]
    C1[TransportLayerClient.download_media_r0]
    C2[TransportLayerClient.download_media_v3]
    D[MatrixFederationHttpClient.get_file]
    A --> D

    A -.-> B -.-> C2 -.-> D
    B -.-> C1 -.-> D
```

Fixes #15196, part of #15661.

### CODE REVIEW

File: synapse/federation/transport/client.py [hunk 1]
```diff
@@ -804,6 +805,58 @@ async def get_account_status(
...
+            args={
+                # tell the remote server to 404 if it doesn't
+                # recognise the server_name, to make sure we don't
+                # end up with a routing loop.
+                "allow_remote": "false",
+                "timeout_ms": str(max_timeout_ms),
+            },
+        )
+
+    async def download_media_v3(
```
Comment (clokep) [inline synapse/federation/transport/client.py:832]: These are very repetitive, but it is the style of this file. 🤷

File: synapse/http/matrixfederationclient.py [hunk 2]
```diff
@@ -714,6 +720,26 @@ async def _send_request(
                             response.code,
                             response_phrase,
                         )
+                    elif (
+                        response.code in (307, 308)
```
Comment (clokep) [inline synapse/http/matrixfederationclient.py:724]: I thought about wrapping the agent in a `RedirectAgent`, but that handles additional codes sadly.

File: synapse/http/matrixfederationclient.py [hunk 3]
```diff
@@ -153,12 +153,18 @@ class MatrixFederationRequest:
...
-    """The URI of this request
+    uri: bytes = b""
+    """The URI of this request, usually generated from the above information.
+    """
+
+    _generate_uri: bool = True
+    """True to automatically generate the uri field based on the above information.
+
+    Set to False if manually configuring the URI.
     """
```
Comment (clokep) [inline synapse/http/matrixfederationclient.py:168]: I couldn't figure out a better way to force attrs to do what I want. Using a default/factory works if you're either just creating the instance or evolving it and updating the URI, but *not* when evolving it and wanting to generate the URI again. I figured being explicit was best.

File: synapse/federation/transport/client.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/federation/transport/client.py:832]: Is there any difference in the impls other than `s/r0/v3/`?

File: tests/media/test_media_storage.py [hunk 4]
```diff
@@ -257,10 +258,15 @@ def write_to(
                 output_stream.write(data)
                 return response
 
+            def write_err(f: Failure) -> Failure:
+                f.trap(HttpResponseException)
+                output_stream.write(f.value.response)
+                return f
```
Comment (DMRobertson) [inline tests/media/test_media_storage.py:265]: This one was new to me: https://docs.twisted.org/en/stable/api/twisted.python.failure.Failure.html#trap

TL;DR the trap call is a no-op if `f ` contains an `HTTPResponseException`; otherwise the trap raises immediately so that the next errback can handle this Failure.

Comment (DMRobertson) [reply tests/media/test_media_storage.py:265]: I'd missed that this was in test code though. Why do we need to add this as an errback all of a sudden?

File: synapse/federation/transport/client.py [hunk 5]
```diff
@@ -850,7 +850,11 @@ async def download_media_v3(
                 # end up with a routing loop.
                 "allow_remote": "false",
                 "timeout_ms": str(max_timeout_ms),
+                # Matix 1.7 allows for this to redirect to another URL, this should
```
Comment (DMRobertson) [inline synapse/federation/transport/client.py:853]:
```suggestion
                # Matrix 1.7 allows for this to redirect to another URL, this should
```

File: synapse/http/matrixfederationclient.py [hunk 6]
```diff
@@ -714,6 +725,26 @@ async def _send_request(
...
+                        response.code in (307, 308)
+                        and follow_redirects
+                        and response.headers.hasHeader("Location")
+                    ):
+                        # The Location header *might* be relative so resolve it.
+                        location = response.headers.getRawHeaders(b"Location")[0]
+                        new_uri = urllib.parse.urljoin(request.uri, location)
+
+                        return await self._send_request(
+                            attr.evolve(request, uri=new_uri, generate_uri=False),
```
Comment (DMRobertson) [inline synapse/http/matrixfederationclient.py:738]: it's spelled `_generate_uri` in the struct definition. Is this some attrs way of declaring a private attribute that you can initialise via the constructor?

File: tests/media/test_media_storage.py (hunk 4 above)
Comment (clokep) [reply tests/media/test_media_storage.py:265]: We need to add it because we know call `errback` sometimes on the list of Deferreds so that we can resolve a request with an error instead of a response.

File: synapse/http/matrixfederationclient.py (hunk 6 above)
Comment (clokep) [reply synapse/http/matrixfederationclient.py:738]: > Is this some attrs way of declaring a private attribute that you can initialise via the constructor?

Yes, pretty much. attrs strips the preceding underscores to let you have 'private' attributes that can be initialized.

File: synapse/federation/transport/client.py (hunk 1 above)
Comment (clokep) [reply synapse/federation/transport/client.py:832]: The additional `allow_redirects` parameter for `v3`.

Comment (DMRobertson) [reply synapse/federation/transport/client.py:832]: Aha, thanks.

File: tests/media/test_media_storage.py (hunk 4 above)
Comment (clokep) [reply tests/media/test_media_storage.py:265]: See https://github.com/matrix-org/synapse/pull/16701/files#diff-50994e96eaadcf21dd111024e283f06c231adc413a3530ce7356e02ada339a3eR704-R706

**Review by DMRobertson — APPROVED** [review]
SGTM, thanks!

File: synapse/http/matrixfederationclient.py (hunk 6 above)
Comment (clokep) [reply synapse/http/matrixfederationclient.py:738]: https://www.attrs.org/en/stable/init.html#private-attributes FTR

File: tests/media/test_media_storage.py (hunk 4 above)
Comment (DMRobertson) [reply tests/media/test_media_storage.py:265]: Ahhh, I think I see: we never called errback on the mock until now!

**Comment by clokep** [comment]
Something like https://github.com/matrix-org/synapse-s3-storage-provider/issues/107 might be useful if you're interested in supporting this for the Client-Server API. Implementing it in Synapse proper seems out of scope?
